sno/working_copy/postgis_adapter.py

```diff
@@ -0,0 +1,280 @@
+    id_salt - the UUIDs of the generated ColumnSchema are deterministic and depend on
+    the name and type of the column, and on this salt.
+    """
+    name = pg_col_info["column_name"]
+    pk_index = pg_col_info["pk_ordinal_position"]
+    if pk_index is not None:
+        pk_index -= 1
+    data_type, extra_type_info = _pg_type_to_v2_type(pg_col_info)
+
+    if data_type == "geometry":
+        data_type, extra_type_info = _pg_type_to_v2_geometry_type(
+            name, pg_geometry_columns, pg_spatial_ref_sys
+        )
+
+    col_id = ColumnSchema.deterministic_id(name, data_type, id_salt)
+    return ColumnSchema(col_id, name, data_type, pk_index, **extra_type_info)
+
+
+def _pg_type_to_v2_type(pg_col_info):
+    v2_type_info = _PG_TYPE_TO_V2_TYPE.get(pg_col_info["data_type"])
+    if v2_type_info is None:
+        v2_type_info = _PG_TYPE_TO_V2_TYPE.get(pg_col_info["udt_name"])
+
+    if isinstance(v2_type_info, tuple):
+        v2_type = v2_type_info[0]
+        extra_type_info = {"size": v2_type_info[1]}
+    else:
+        v2_type = v2_type_info
+        extra_type_info = {}
+
+    # TODO: standardise on null vs not-present for extra_type_info.
+    # TODO: Fix legacy problems caused by any inconsistency.
+    if v2_type == "text":
+        length = pg_col_info["character_maximum_length"] or None
+        if length is not None:
+            extra_type_info["length"] = length
+
+    if v2_type == "numeric":
+        extra_type_info["precision"] = pg_col_info["numeric_precision"] or None
+        extra_type_info["scale"] = pg_col_info["numeric_precision"] or None
+
+    return v2_type, extra_type_info
+
+
+def _pg_type_to_v2_geometry_type(col_name, pggc, pgsrs):
+    """
+    col_name - the name of the column.
+    pggc - the contents of the 'geometry_columns' table.
+    pgsrs - the contents of the 'spatial_ref_sys' table.
+    """
+    geom_col_info = next((r for r in pggc if r["f_geometry_column"] == col_name))
+
+    geometry_type = geom_col_info["type"].upper()
+    # Look for Z, M, or ZM suffix
+    geometry_type, m = _pop_suffix(geometry_type, "M")
+    geometry_type, z = _pop_suffix(geometry_type, "Z")
+    geometry_type = f"{geometry_type} {z}{m}".strip()
+
+    geometry_crs = None
+    crs_id = geom_col_info["srid"]
+    if crs_id:
+        crs_info = next((r for r in pgsrs if r["srid"] == crs_id), None)
+        if crs_info:
+            geometry_crs = crs_util.get_identifier_str(crs_info["srtext"])
+
+    return "geometry", {"geometryType": geometry_type, "geometryCRS": geometry_crs}
+
+
+def _pop_suffix(geometry_type, suffix):
+    """
+    Returns (geometry-type-without-suffix, suffix) if geometry-type ends with suffix.
+    Otherwise just returns (geometry-type, "")
+    """
+    if geometry_type.endswith(suffix):
+        return geometry_type[:-1], suffix
+    else:
+        return geometry_type, ""
+
+
+def generate_postgis_geometry_columns(v2_obj, postgis_schema):
```

Comment:
we need to update the column definitions via `ALTER TABLE`, not update `geometry_columns`

Reply:
Done (as much as is required for this change).

The scope of this change only supports diffs and commits - it doesn't yet support resets. So, the ALTER TABLE only happens when the user changes the table, or in a test. So, I haven't exactly "done" this comment, but I will in the reset PR.

On the other hand, I no longer create my own `geometry_columns` table - the code now only ever reads `public.geometry_columns` and never writes to it. So, that is fixed.